cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -1127,6 +1127,66 @@ class ZeroCopyCast : public CastKernelBase {
   }
 };
 
+class ExtensionCastKernel : public CastKernelBase {
+ public:
+  static Status Make(const DataType& in_type, std::shared_ptr<DataType> out_type,
+                     const CastOptions& options,
+                     std::unique_ptr<CastKernelBase>* kernel) {
+    if (in_type.id() != Type::EXTENSION) {
+      return Status::TypeError("Not an extension type");
+    }
```

Comment:
I don't think this check is needed if this "can't happen" -- the checked_cast should be sufficient

Reply:
Removed